smithy-aws-endpoints/src/main/java/software/amazon/smithy/rulesengine/aws/language/functions/endpoints/StandardPartitionalEndpointsTrait.java

```diff
@@ -0,0 +1,148 @@
+    public static final ShapeId ID = ShapeId.from("aws.endpoints#standardPartitionalEndpoints");
+    public static final String PARTITION_ENDPOINT_SPECIAL_CASES = "partitionEndpointSpecialCases";
+    public static final String ENDPOINT_PATTERN_TYPE = "endpointPatternType";
+
+    private final Map<String, List<PartitionEndpointSpecialCase>> partitionEndpointSpecialCases;
+
+    private final EndpointPatternType endpointPatternType;
+
+    public StandardPartitionalEndpointsTrait(StandardPartitionalEndpointsTrait.Builder builder) {
+        super(ID, builder.getSourceLocation());
+        partitionEndpointSpecialCases = builder.partitionEndpointSpecialCases.copy();
+        endpointPatternType = Objects.requireNonNull(builder.endpointPatternType);
+    }
+
+    @Override
+    protected Node createNode() {
+        final Node partitionEndpointSpecialCasesNode = partitionEndpointSpecialCases.entrySet().stream()
+            .map(entry -> {
+                List<Node> nodes = new ArrayList<>();
+                for (PartitionEndpointSpecialCase partitionEndpointSpecialCase : entry.getValue()) {
+                    nodes.add(partitionEndpointSpecialCase.toNode());
+                }
+                return new AbstractMap.SimpleImmutableEntry<>(
+                    Node.from(entry.getKey()), Node.fromNodes(nodes));
+            })
+            .collect(ObjectNode.collect(Map.Entry::getKey, Map.Entry::getValue));
+
+        return ObjectNode.objectNodeBuilder()
+            .sourceLocation(getSourceLocation())
+            .withMember(PARTITION_ENDPOINT_SPECIAL_CASES, partitionEndpointSpecialCasesNode)
+            .withMember(ENDPOINT_PATTERN_TYPE, endpointPatternType.getName())
+            .build();
+    }
+
+    @Override
+    public SmithyBuilder<StandardPartitionalEndpointsTrait> toBuilder() {
+        return new Builder()
+                .partitionEndpointSpecialCases(partitionEndpointSpecialCases)
+                .endpointPatternType(endpointPatternType);
+    }
+
+    public Map<String, List<PartitionEndpointSpecialCase>> getPartitionEndpointSpecialCases() {
+        return partitionEndpointSpecialCases;
+    }
+
+    public EndpointPatternType getEndpointPatternType() {
+        return endpointPatternType;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Provider extends AbstractTrait.Provider {
+        public Provider() {
+            super(ID);
+        }
+
+        @Override
+        public Trait createTrait(ShapeId target, Node value) {
+            ObjectNode objectNode = value.expectObjectNode();
+
+            EndpointPatternType endpointPatternType = EndpointPatternType
+                .fromNode(objectNode.expectStringMember(ENDPOINT_PATTERN_TYPE));
+
+            StandardPartitionalEndpointsTrait.Builder builder = builder()
+                    .sourceLocation(value)
+                    .endpointPatternType(endpointPatternType);
+
+            objectNode
```

Comment:
This would be easier to read if it was done with a conditional to check for the presence of this member, then looping over its members and the nested array node.